vm/src/deserializer.h

```diff
@@ -0,0 +1,83 @@
+#ifndef BORGES_DESERIALIZER_H
+#define BORGES_DESERIALIZER_H
+
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace Borges {
+
+class Deserializer {
+public:
+    Deserializer(std::size_t n, const double* p);
+    void deserialize(double* p);
```

Comment:
why is the deserialize with (double *p) a separate function (and not implemented via the shared_ptr variant ?)

Reply:
This is a good question, and I can see why this would be non obvious. To answer this question, I need to establish some context first, so please bear with me.

Suppose we want to be able to deserialize values like this:

    std::vector<double> xs = deserialize<std::vector<double>>(d)

To achieve this, we could define deserialize as a template function, and then specialize this template for each deserializable type. For instance, we can write:

    template <typename T> T deserialize(Deserializer& d);

If we then want to specialise that function for different types, we would write:

    template <> bool deserialize<bool>(Deserializer& d);
    template <> bool deserialize<double>(Deserializer& d);

So far, so good.

Unfortunately, the above approach breaks down when we want to deserialize more complex types. The first problem is that C++ does not allow partial specialization of template functions. That is, we cannot specialize deserialize for, say, vectors of an arbitrary type:

    template <typename T>
    std::vector<T> deserialize<std:vector<T>>(Deserializer& d); // THIS IS NOT ALLOWED!

The second problem is that to deserialize a value of type T, the deserialize function needs to create a temporary variable tmp of type T, into which the value will be read. This is not a problem for POD-types like bool or double, but doesn't work for types that don't have a default constructor, such as Curve. In that case, writing T tmp; will lead to a compiler error.

We could require that each serializable type needs to have a default constructor, but that's an artificial restriction, and still doesn't solve the problem of partial template function specialization. Instead, we will tackle both the above problems at the same time, using the following approach:

Instead of creating a temporary value of type T, deserialize creates a buffer of uninitialized memory thats large enough to contain a value of type T (this buffer needs to be properly aligned, so we use the alignof operator to achieve this). We then use operator new() to directly construct a value of type T in this uninitialized memory, using the appropriate constructor. By directly constructing the value in uninitialized memory instead of assigning it to a temporary value, we avoid a needless call to the (possibly non-existent) default constructor.

To solve the problem of partial template function specialization, we introduce a level of indirection: instead of initialising the above pointer directly, deserialize calls a second deserialize function that takes the uninitialized pointer as an argument. This second function is responsible for doing the actual deserialization and constructing the value. Crucially, because this second function is *not* a template, it can be freely overloaded for any type T we want.

    template <typename T>
    std::vector<T> deserialize(Deserializer& d, std::vector<double>* p); // THIS IS ALLOWED!

Reply:
Now that I've explained the context of what we're trying to do, let me get back to your original question. What you are looking at is a member function that deserializes and constructs a double, using the given pointer. In contrast, the shared_ptr variant deserializes and constructs a shared_ptr, using the given pointer.

As an aside, these are the only two functions that need direct access to the state of the Deserializer, which is why they are member functions. All the other functions can be implemented in terms of deserializing doubles or shared_ptrs.

Reply:
ok